src/Type/Schema.php

```diff
@@ -353,15 +355,18 @@ private function loadType($typeName)
 
         $type = $typeLoader($typeName);
 
-        if (! $type instanceof Type) {
+        if (! $type instanceof Type && ! is_callable($type)) {
             throw new InvariantViolation(
                 sprintf(
-                    'Type loader is expected to return valid type "%s", but it returned %s',
+                    'Type loader is expected to return a callable or valid type "%s", but it returned %s',
                     $typeName,
                     Utils::printSafe($type)
                 )
             );
         }
+
+        $type = Type::resolveLazyType($type);
```

Comment:
Not sure if we should allow `typeLoader` to return `callable`. It is basically useless because when it's called - we already need the actual type. Do you have any reasons to allow it?

Reply:
Sorry, I don't understand--what are you suggesting?

Reply:
I suggest disallowing `typeLoader` to return `callable`. It should always return a specific type.

Reply:
I understand what you mean @vladar

There are two levels of lazy loading going on here:

1. Calling `$typeLoader($typeName)`
2. Checking if the result of that is yet another callable and calling that again

Why would 2. be necessary? As you noted, the two calls happen right after each other.

Reply:
Is this what you guys had in mind? 484d2dd

I dunno, I guess I'm okay with it, but I kind of like the idea that executing a lazy type is entirely the library's responsibility, and all the client has to do is deliver a callable or a type, but I suppose I can go either way. Let me know what you think.

Reply:
@shmax thanks for showing the impact of changing the approach in that commit.

I can see how it can more convenient to implement if the type loading is lazy in all circumstances. It allows using the same lazy type reference in any place that expects a type, including the type loader.

The performance impact should be pretty minimal, there is one extra call to check `is_callable()` for every type loaded during a request. I think the extra convenience in development and the simplicity are worth it.

@vladar can you make a final call?

Reply:
Yeah, I see how this improves DX, let's restore it in a follow-up PR!